spec/models/orchestration_stack/retirement_management_spec.rb

```diff
@@ -9,14 +11,37 @@
       @stack = FactoryBot.create(:orchestration_stack, :ext_management_system => ems)
     end
 
-    it "#retirement_check" do
-      expect(MiqEvent).to receive(:raise_evm_event)
-      @stack.update_attributes(:retires_on => 90.days.ago, :retirement_warn => 60, :retirement_last_warn => nil)
-      expect(@stack.retirement_last_warn).to be_nil
-      @stack.retirement_check
-      @stack.reload
-      expect(@stack.retirement_last_warn).not_to be_nil
-      expect(@stack.retirement_requester).to eq(user.userid)
+    describe "#retirement_check" do
+      context "with user" do
+        it "uses user as requester" do
+          expect(MiqEvent).to receive(:raise_evm_event)
+          stack_with_owner.update_attributes(:retires_on => 90.days.ago, :retirement_warn => 60, :retirement_last_warn => nil)
+          expect(stack_with_owner.retirement_last_warn).to be_nil
+          stack_with_owner.retirement_check
+          stack_with_owner.reload
+          expect(stack_with_owner.retirement_last_warn).not_to be_nil
+          expect(stack_with_owner.retirement_requester).to eq(user.userid)
+        end
+      end
+
+      context "with deleted user" do
+        before do
+          # system_context_retirement relies on the presence of a user with this userid
+          FactoryBot.create(:user, :userid => 'admin', :role => 'super_administrator')
+          user.destroy
+          stack_with_owner.reload
+        end
+
+        it "uses admin as requester" do
+          expect(MiqEvent).to receive(:raise_evm_event)
+          stack_with_owner.update_attributes(:retires_on => 90.days.ago, :retirement_warn => 60, :retirement_last_warn => nil)
+          expect(stack_with_owner.retirement_last_warn).to be_nil
+          stack_with_owner.retirement_check
+          stack_with_owner.reload
+          expect(stack_with_owner.retirement_last_warn).not_to be_nil
+          expect(stack_with_owner.retirement_requester).to eq("admin")
```

Comment:
Interesting, I would not have expected that.  Is this okay for audit purposes?

Reply:
retirement_check only runs in the system_context -- in which someone scheduled a retirement for a future date. Our modus operandi at the moment for system_context is always to set the requester to admin.